gobblin-core/src/main/java/gobblin/source/extractor/extract/google/GoogleDriveFileSystem.java

```diff
@@ -0,0 +1,311 @@
+    String pageToken = null;
+    List<File> result = new ArrayList<>();
+    Optional<String> query = buildQuery(folderId, fileId);
+
+    do {
+      Drive.Files.List request = client.files()
+          .list()
+          .setFields("files/id,files/mimeType,files/modifiedTime,files/size,files/permissions")
+          .setPageSize(pageSize);
+      if (query.isPresent()) {
+        request = request.setQ(query.get());
+      }
+
+      if (pageToken != null) {
+        request = request.setPageToken(pageToken);
+      }
+
+      LOG.info("Google drive List request: " + request);
+      if (LOG.isDebugEnabled()) {
+        LOG.debug("Google drive List request: " + request);
+      }
+
+      FileList fileList = null;
+      try {
+        fileList = request.execute();
+      } catch (GoogleJsonResponseException e) {
+        GoogleJsonError error = e.getDetails();
+        if (404 == error.getCode()) {
+          throw new FileNotFoundException("File not found. Request: " + request);
+        }
+        throw e;
+      }
+
+      pageToken = fileList.getNextPageToken();
+
+      List<File> files = fileList.getFiles();
+      if (files == null || files.isEmpty()) {
+        return result;
+      }
+      result.addAll(files);
+    } while (pageToken != null);
+    return result;
+  }
+
+  /**
+   * Build query for Google drive.
+   * @see https://developers.google.com/drive/v3/web/search-parameters
+   *
+   * @param folderId
+   * @param fileName
+   * @return Query
+   */
+  @VisibleForTesting
+  Optional<String> buildQuery(String folderId, String fileName) {
+    if (StringUtils.isEmpty(folderId) && StringUtils.isEmpty(fileName)) {
+      return Optional.absent();
+    }
+
+    StringBuilder query = new StringBuilder();
+    if (StringUtils.isNotEmpty(folderId)) {
+      query.append("'").append(folderId).append("'")
+           .append(" in parents");
+    }
+    if (StringUtils.isNotEmpty(fileName)) {
+      if (query.length() > 0) {
+        query.append(" and ");
+      }
+      query.append("name contains ")
+           .append("'").append(fileName).append("'");
+    }
+    return Optional.of(query.toString());
+  }
+
+  /**
+   * org.apache.hadoop.fs.Path assumes that there separator in file system naming and "/" is the separator.
+   * When org.apache.hadoop.fs.Path sees "/" in path String, it splits into parent and name. As fileID is a random
+   * String determined by Google and it can contain "/" itself, this method check if parent and name is separated and
+   * restore "/" back to file ID.
+   *
+   * @param p
+   * @return
+   */
+  public static String toFileId(Path p) {
+    if (p.isRoot()) {
+      return "";
+    }
+    final String format = "%s" + Path.SEPARATOR + "%s";
+    if (p.getParent() != null && StringUtils.isEmpty(p.getParent().getName())) {
+      return p.getName();
+    }
+    return String.format(format, toFileId(p.getParent()), p.getName());
```

Comment:
What if there are multiple / in the id, you probably need to recursively call toFileId()